packages/portal-contracts/contracts/messaging/gateway/FuelERC721Gateway.sol

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.9;
+
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+import {UUPSUpgradeable} from "@openzeppelin/contracts-upgradeable/proxy/utils/UUPSUpgradeable.sol";
+import {AccessControlUpgradeable} from "@openzeppelin/contracts-upgradeable/access/AccessControlUpgradeable.sol";
+import {PausableUpgradeable} from "@openzeppelin/contracts-upgradeable/security/PausableUpgradeable.sol";
+import {IERC721Upgradeable} from "@openzeppelin/contracts-upgradeable/token/ERC721/IERC721Upgradeable.sol";
+import {FuelBridgeBase} from "./FuelBridgeBase.sol";
+import {FuelMessagePortal, CommonPredicates} from "../../fuelchain/FuelMessagePortal.sol";
+import {FuelMessagesEnabledUpgradeable} from "../FuelMessagesEnabledUpgradeable.sol";
+
+import "hardhat/console.sol";
+
+/// @title FuelERC721Gateway
+/// @notice The L1 side of the general ERC721 gateway with Fuel
+/// @dev This contract can be used as a template for future gateways to Fuel
+contract FuelERC721Gateway is
+    Initializable,
+    FuelBridgeBase,
+    FuelMessagesEnabledUpgradeable,
+    PausableUpgradeable,
+    AccessControlUpgradeable,
+    UUPSUpgradeable
+{
+    ////////////
+    // Events //
+    ////////////
+
+    /// @dev Emitted when tokens are deposited from Ethereum to Fuel
+    event Deposit(bytes32 indexed sender, address indexed tokenAddress, bytes32 fuelContractId, uint256 tokenId);
```

Comment:
Wondering if we want to easily index via the contract id too?
Similarly, in withdrawal below.

Reply:
OK for me - I 'd do it in the ERC20 gateway too for the sake of consistency

Reply:
Done